File: packages/node-agent/src/toolRuntime.ts
import { readOnlyTools, toolNames } from './catalog.js';
import type { JsonObject } from './types.js';

export type ToolDomain =
  | 'harness'
  | 'runtime'
  | 'history'
  | 'task'
  | 'filesystem'
  | 'search'
  | 'quality'
  | 'process'
  | 'git'
  | 'desktop';

export type ToolExecutionLane = 'blocking' | 'process' | 'control';
export type ToolLockGroup = 'history' | 'workspace_content' | 'git' | 'task' | 'cwd';
export type ToolPermissionKind = 'workspace_mutation' | 'process_execution' | 'network' | 'privileged_operation';
export type ToolUsageFamily = 'filesystem' | 'search' | 'quality' | 'process' | 'git' | 'history' | 'runtime' | 'other';
export type ToolCoalescingPolicy = 'never' | 'read_only' | 'operation_id';
export type ToolMutationPolicy = 'never' | 'always' | 'format_apply';

export interface ToolRuntimeDescriptor {
  readonly name: string;
  readonly canonicalName: string;
  readonly domain: ToolDomain;
  readonly usageFamily: ToolUsageFamily;
  readonly lane: ToolExecutionLane;
  readonly lockGroups: readonly ToolLockGroup[];
  readonly harnessTool: boolean;
  readonly guardedPermission?: ToolPermissionKind;
  readonly coalescing: ToolCoalescingPolicy;
  readonly mutation: ToolMutationPolicy;
  readonly workspaceSelector: boolean;
}

type ToolRuntimeOverrides = Partial<Pick<ToolRuntimeDescriptor,
  'usageFamily' | 'lane' | 'lockGroups' | 'harnessTool' | 'guardedPermission' | 'coalescing' | 'mutation' | 'workspaceSelector'
>>;

interface ToolRuntimeModule {
  readonly domain: ToolDomain;
  readonly usageFamily: ToolUsageFamily;
  readonly defaults?: ToolRuntimeOverrides;
  readonly tools: Readonly<Record<string, ToolRuntimeOverrides>>;
}

const WORKSPACE_CONTENT_LOCK = ['workspace_content'] as const;
const GIT_LOCK = ['git'] as const;
const TASK_LOCK = ['task'] as const;

function workspaceSelectorFor(name: string): boolean {
  return name.startsWith('git_') || [
    'set_default_cwd',
    'read_file', 'read_many', 'list_files', 'project_map', 'search_text',
    'apply_patch', 'edit', 'file_ops', 'patch_check', 'format_files', 'view_image',
    'exec_health_check', 'exec_command', 'exec_many', 'wait_command', 'resolve_operation',
    'list_sessions', 'send_input', 'kill_session', 'read_output', 'request_permissions'
  ].includes(name);
}

const TOOL_RUNTIME_MODULES: readonly ToolRuntimeModule[] = [
  {
    domain: 'harness',
    usageFamily: 'other',
    defaults: { harnessTool: true },
    tools: {
      harness_status: {},
      operation_log: {}
    }
  },
  {
    domain: 'runtime',
    usageFamily: 'other',
    tools: {
      server_info: { usageFamily: 'runtime' },
      list_workspace_folders: {},
      conversation_bootstrap: { lockGroups: ['history'], coalescing: 'never' },
      switch_workspace_folder: { coalescing: 'never' },
      query_tool_usage: { usageFamily: 'runtime' },
      set_default_cwd: { usageFamily: 'runtime', lockGroups: ['cwd'] },
      request_permissions: { usageFamily: 'runtime', lane: 'control', coalescing: 'never' }
    }
  },
  {
    domain: 'history',
    usageFamily: 'history',
    defaults: { lockGroups: ['history'] },
    tools: {
      history_session_bootstrap: {},
      history_session_checkpoint: {},
      history_session_validate: {}
    }
  },
  {
    domain: 'task',
    usageFamily: 'other',
    defaults: { harnessTool: true },
    tools: {
      project_state: {},
      start_task: { lockGroups: TASK_LOCK, mutation: 'always' },
      update_task: { lockGroups: TASK_LOCK, mutation: 'always' },
      pause_task: { lockGroups: TASK_LOCK, mutation: 'always' },
      resume_task: { lockGroups: TASK_LOCK, mutation: 'always' },
      finish_task: { lockGroups: TASK_LOCK, mutation: 'always' },
      task_context: {},
      list_task_events: {},
      change_summary: {}
    }
  },
  {
    domain: 'filesystem',
    usageFamily: 'filesystem',
    tools: {
      read_file: {},
      read_many: {},
      list_files: {},
      apply_patch: {
        lockGroups: WORKSPACE_CONTENT_LOCK,
        guardedPermission: 'workspace_mutation',
        mutation: 'always'
      },
      edit: {
        lockGroups: WORKSPACE_CONTENT_LOCK,
        guardedPermission: 'workspace_mutation',
        mutation: 'always'
      },
      file_ops: {
        lockGroups: WORKSPACE_CONTENT_LOCK,
        guardedPermission: 'workspace_mutation',
        mutation: 'always'
      },
      patch_check: { usageFamily: 'other' },
      view_image: {}
    }
  },
  {
    domain: 'desktop',
    usageFamily: 'runtime',
    tools: {
      desktop_displays: {},
      desktop_screenshot: {},
      desktop_click: { mutation: 'always' },
      desktop_drag: { mutation: 'always' },
      desktop_scroll: { mutation: 'always' },
      desktop_type: { mutation: 'always' },
      desktop_key: { mutation: 'always' }
    }
  },
  {
    domain: 'search',
    usageFamily: 'search',
    tools: {
      project_map: {},
      search_text: {}
    }
  },
  {
    domain: 'quality',
    usageFamily: 'quality',
    tools: {
      format_files: {
        lockGroups: WORKSPACE_CONTENT_LOCK,
        guardedPermission: 'privileged_operation',
        mutation: 'format_apply'
      }
    }
  },
  {
    domain: 'process',
    usageFamily: 'process',
    tools: {
      exec_health_check: { usageFamily: 'other' },
      exec_command: {
        lane: 'process',
        guardedPermission: 'process_execution',
        coalescing: 'operation_id',
        mutation: 'always'
      },
      exec_many: {
        lane: 'process',
        guardedPermission: 'process_execution',
        mutation: 'always'
      },
      wait_command: { lane: 'control' },
      resolve_operation: { lane: 'control' },
      list_sessions: { lane: 'control' },
      send_input: { lane: 'control' },
      kill_session: {
        lane: 'control',
        guardedPermission: 'process_execution',
        mutation: 'always'
      },
      read_output: { lane: 'control' }
    }
  },
  {
    domain: 'git',
    usageFamily: 'git',
    tools: {
      git_status: {},
      git_diff: {},
      git_log: {},
      git_show: {},
      git_blame: {},
      git_branch: {
        lockGroups: GIT_LOCK,
        guardedPermission: 'workspace_mutation',
        mutation: 'always'
      },
      git_worktree: {
        lockGroups: ['git', 'workspace_content'],
        guardedPermission: 'workspace_mutation',
        mutation: 'always'
      },
      git_stage: {
        lockGroups: GIT_LOCK,
        guardedPermission: 'workspace_mutation',
        mutation: 'always'
      },
      git_commit: {
        lockGroups: GIT_LOCK,
        guardedPermission: 'workspace_mutation',
        mutation: 'always'
      },
      git_push: { guardedPermission: 'network' },
      git_restore: {
        lockGroups: ['git', 'workspace_content'],
        guardedPermission: 'workspace_mutation',
        mutation: 'always'
      }
    }
  }
];

const runtimeByName = new Map<string, ToolRuntimeDescriptor>();

for (const moduleDefinition of TOOL_RUNTIME_MODULES) {
  for (const [name, overrides] of Object.entries(moduleDefinition.tools)) {
    if (runtimeByName.has(name)) throw new Error(`Duplicate Node tool runtime metadata: ${name}`);
    const configured = { ...moduleDefinition.defaults, ...overrides };
    runtimeByName.set(name, Object.freeze({
      name,
      canonicalName: name,
      domain: moduleDefinition.domain,
      usageFamily: configured.usageFamily ?? moduleDefinition.usageFamily,
      lane: configured.lane ?? 'blocking',
      lockGroups: Object.freeze([...(configured.lockGroups ?? [])]),
      harnessTool: configured.harnessTool ?? false,
      guardedPermission: configured.guardedPermission,
      coalescing: configured.coalescing ?? (readOnlyTools.has(name) ? 'read_only' : 'never'),
      mutation: configured.mutation ?? 'never',
      workspaceSelector: configured.workspaceSelector ?? workspaceSelectorFor(name)
    }));
  }
}

const catalogNames = new Set(toolNames);
const missingRuntime = toolNames.filter(name => !runtimeByName.has(name));
const unknownRuntime = [...runtimeByName.keys()].filter(name => !catalogNames.has(name));
if (missingRuntime.length || unknownRuntime.length) {
  throw new Error([
    missingRuntime.length ? `Missing Node tool runtime metadata: ${missingRuntime.join(', ')}` : '',
    unknownRuntime.length ? `Unknown Node tool runtime metadata: ${unknownRuntime.join(', ')}` : ''
  ].filter(Boolean).join('; '));
}

for (const alias of ['edit_file', 'edit_many']) {
  const edit = runtimeByName.get('edit');
  if (!edit) throw new Error('Node tool runtime alias target is missing: edit');
  runtimeByName.set(alias, Object.freeze({ ...edit, name: alias, canonicalName: 'edit' }));
}

const unknownRuntimeDefaults = Object.freeze({
  domain: 'runtime' as const,
  usageFamily: 'other' as const,
  lane: 'blocking' as const,
  lockGroups: Object.freeze([]) as readonly ToolLockGroup[],
  harnessTool: false,
  coalescing: 'never' as const,
  mutation: 'never' as const,
  workspaceSelector: false
});

export function registeredToolRuntimeNames(): string[] {
  return toolNames.filter(name => runtimeByName.has(name));
}

export function toolRuntimeFor(name: string): ToolRuntimeDescriptor {
  return runtimeByName.get(name) ?? {
    ...unknownRuntimeDefaults,
    name,
    canonicalName: name
  };
}

export function canCoalesceToolCall(name: string, args: JsonObject): boolean {
  const policy = toolRuntimeFor(name).coalescing;
  if (policy === 'read_only') return true;
  return policy === 'operation_id' && String(args.operation_id ?? '').trim().length > 0;
}

export function canonicalToolCall(name: string, args: JsonObject): { name: string; args: JsonObject } {
  const canonicalName = toolRuntimeFor(name).canonicalName;
  if (name === 'edit_file') {
    const file: JsonObject = {};
    for (const field of ['path', 'expected_sha256', 'edits', 'apply_proposal']) {
      if (args[field] !== undefined) file[field] = args[field];
    }
    const converted: JsonObject = { files: [file] };
    if (args.dry_run !== undefined) converted.dry_run = args.dry_run;
    if (args.reason !== undefined) converted.reason = args.reason;
    return { name: canonicalName, args: converted };
  }
  return { name: canonicalName, args };
}

export function requestMutates(tool: string, args: JsonObject): boolean {
  const policy = toolRuntimeFor(tool).mutation;
  if (policy === 'always') return true;
  if (policy === 'format_apply') return String(args.mode ?? '') === 'apply';
  return false;
}

export function toolUsageFamily(tool: string): ToolUsageFamily {
  return toolRuntimeFor(tool).usageFamily;
}

File: src-tauri/src/tools/tool_runtime.rs
use serde_json::Value;

#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd)]
pub(crate) enum MutationLockGroup {
    History,
    WorkspaceContent,
    Git,
    Task,
    Cwd,
}

impl MutationLockGroup {
    pub(crate) fn as_str(self) -> &'static str {
        match self {
            Self::History => "history",
            Self::WorkspaceContent => "workspace_content",
            Self::Git => "git",
            Self::Task => "task",
            Self::Cwd => "cwd",
        }
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub(crate) enum ToolExecutionLane {
    Fast,
    Blocking,
    Process,
    Control,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub(crate) enum ToolMutationPolicy {
    Never,
    Always,
    FormatApply,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub(crate) enum ToolBaselinePolicy {
    Never,
    Always,
    UnlessDryRun,
    FormatApply,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub(crate) enum ToolUsageFamily {
    Filesystem,
    Search,
    Quality,
    Process,
    Git,
    History,
    Runtime,
    Other,
}

impl ToolUsageFamily {
    pub(crate) fn as_str(self) -> &'static str {
        match self {
            Self::Filesystem => "filesystem",
            Self::Search => "search",
            Self::Quality => "quality",
            Self::Process => "process",
            Self::Git => "git",
            Self::History => "history",
            Self::Runtime => "runtime",
            Self::Other => "other",
        }
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub(crate) struct ToolRuntimeDescriptor {
    pub lane: ToolExecutionLane,
    pub lock_groups: &'static [MutationLockGroup],
    pub mutation: ToolMutationPolicy,
    pub baseline: ToolBaselinePolicy,
    pub standalone_operation: bool,
    pub log_operation: bool,
    pub usage_family: ToolUsageFamily,
    pub workspace_selector: bool,
}

const NO_LOCKS: &[MutationLockGroup] = &[];
const HISTORY_LOCK: &[MutationLockGroup] = &[MutationLockGroup::History];
const WORKSPACE_CONTENT_LOCK: &[MutationLockGroup] = &[MutationLockGroup::WorkspaceContent];
const GIT_LOCK: &[MutationLockGroup] = &[MutationLockGroup::Git];
const GIT_AND_WORKSPACE_LOCK: &[MutationLockGroup] =
    &[MutationLockGroup::WorkspaceContent, MutationLockGroup::Git];
const TASK_LOCK: &[MutationLockGroup] = &[MutationLockGroup::Task];
const CWD_LOCK: &[MutationLockGroup] = &[MutationLockGroup::Cwd];

const DEFAULT: ToolRuntimeDescriptor = ToolRuntimeDescriptor {
    lane: ToolExecutionLane::Blocking,
    lock_groups: NO_LOCKS,
    mutation: ToolMutationPolicy::Never,
    baseline: ToolBaselinePolicy::Never,
    standalone_operation: false,
    log_operation: false,
    usage_family: ToolUsageFamily::Other,
    workspace_selector: false,
};

pub(crate) fn descriptor(name: &str) -> ToolRuntimeDescriptor {
    let canonical = match name {
        "edit_file" | "edit_many" => "edit",
        _ => name,
    };
    let mut runtime = DEFAULT;

    runtime.usage_family = if canonical.starts_with("git_") {
        ToolUsageFamily::Git
    } else if canonical.starts_with("history_session_") {
        ToolUsageFamily::History
    } else {
        match canonical {
            "read_file" | "read_many" | "list_files" | "apply_patch" | "edit" | "file_ops"
            | "view_image" => ToolUsageFamily::Filesystem,
            "search_text" | "project_map" => ToolUsageFamily::Search,
            "format_files" => ToolUsageFamily::Quality,
            "exec_command" | "exec_many" | "wait_command" | "resolve_operation"
            | "list_sessions" | "send_input" | "kill_session" | "read_output" => {
                ToolUsageFamily::Process
            }
            "server_info" | "query_tool_usage" | "set_default_cwd" | "request_permissions" => {
                ToolUsageFamily::Runtime
            }
            "desktop_displays" | "desktop_screenshot" | "desktop_click" | "desktop_drag"
            | "desktop_scroll" | "desktop_type" | "desktop_key" => ToolUsageFamily::Runtime,
            _ => ToolUsageFamily::Other,
        }
    };

    runtime.lane = match canonical {
        "server_info" => ToolExecutionLane::Fast,
        "wait_command" | "resolve_operation" | "list_sessions" | "send_input" | "kill_session"
        | "read_output" => ToolExecutionLane::Control,
        "exec_command" | "exec_many" | "exec_health_check" | "request_permissions" => {
            ToolExecutionLane::Process
        }
        _ => ToolExecutionLane::Blocking,
    };

    runtime.lock_groups = match canonical {
        "conversation_bootstrap"
        | "history_session_bootstrap"
        | "history_session_checkpoint"
        | "history_session_validate" => HISTORY_LOCK,
        "apply_patch" | "edit" | "file_ops" | "format_files" => WORKSPACE_CONTENT_LOCK,
        "git_restore" | "git_worktree" => GIT_AND_WORKSPACE_LOCK,
        "git_branch" | "git_stage" | "git_commit" | "git_push" => GIT_LOCK,
        "start_task" | "update_task" | "pause_task" | "resume_task" | "finish_task" => TASK_LOCK,
        "set_default_cwd" => CWD_LOCK,
        _ => NO_LOCKS,
    };

    runtime.mutation = match canonical {
        "format_files" => ToolMutationPolicy::FormatApply,
        "conversation_bootstrap"
        | "switch_workspace_folder"
        | "history_session_bootstrap"
        | "history_session_checkpoint"
        | "history_session_validate"
        | "apply_patch"
        | "edit"
        | "file_ops"
        | "git_branch"
        | "git_worktree"
        | "git_stage"
        | "git_commit"
        | "git_push"
        | "git_restore"
        | "exec_command"
        | "exec_many"
        | "send_input"
        | "kill_session"
        | "request_permissions"
        | "set_default_cwd"
        | "start_task"
        | "update_task"
        | "pause_task"
        | "resume_task"
        | "finish_task" => ToolMutationPolicy::Always,
        "desktop_click" | "desktop_drag" | "desktop_scroll" | "desktop_type" | "desktop_key" => {
            ToolMutationPolicy::Always
        }
        _ => ToolMutationPolicy::Never,
    };

    runtime.baseline = match canonical {
        "exec_command" => ToolBaselinePolicy::Always,
        "apply_patch" | "edit" | "file_ops" | "git_branch" | "git_worktree" | "git_stage"
        | "git_commit" | "git_restore" => ToolBaselinePolicy::UnlessDryRun,
        "format_files" => ToolBaselinePolicy::FormatApply,
        _ => ToolBaselinePolicy::Never,
    };

    runtime.workspace_selector = canonical.starts_with("git_")
        || matches!(
            canonical,
            "set_default_cwd"
                | "read_file"
                | "read_many"
                | "list_files"
                | "project_map"
                | "search_text"
                | "apply_patch"
                | "edit"
                | "file_ops"
                | "patch_check"
                | "format_files"
                | "view_image"
                | "exec_health_check"
                | "exec_command"
                | "exec_many"
                | "wait_command"
                | "resolve_operation"
                | "list_sessions"
                | "send_input"
                | "kill_session"
                | "read_output"
                | "request_permissions"
        );

    runtime.standalone_operation = matches!(
        canonical,
        "patch_check"
            | "apply_patch"
            | "edit"
            | "file_ops"
            | "format_files"
            | "exec_command"
            | "git_branch"
            | "git_worktree"
            | "git_stage"
            | "git_commit"
            | "git_push"
            | "git_restore"
    );
    runtime.log_operation = runtime.standalone_operation || canonical.starts_with("git_");
    runtime
}

pub(crate) fn request_mutates(name: &str, args: &Value) -> bool {
    match descriptor(name).mutation {
        ToolMutationPolicy::Never => false,
        ToolMutationPolicy::Always => true,
        ToolMutationPolicy::FormatApply => {
            args.get("mode").and_then(Value::as_str) == Some("apply")
        }
    }
}

pub(crate) fn is_mutating_tool(name: &str) -> bool {
    descriptor(name).mutation != ToolMutationPolicy::Never
}

pub(crate) fn requires_write_baseline(name: &str, args: &Value) -> bool {
    match descriptor(name).baseline {
        ToolBaselinePolicy::Never => false,
        ToolBaselinePolicy::Always => true,
        ToolBaselinePolicy::UnlessDryRun => !args
            .get("dry_run")
            .and_then(Value::as_bool)
            .unwrap_or(false),
        ToolBaselinePolicy::FormatApply => {
            args.get("mode").and_then(Value::as_str) == Some("apply")
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn runtime_metadata_covers_critical_dispatch_contracts() {
        assert_eq!(descriptor("server_info").lane, ToolExecutionLane::Fast);
        assert_eq!(descriptor("exec_command").lane, ToolExecutionLane::Process);
        assert_eq!(descriptor("wait_command").lane, ToolExecutionLane::Control);
        assert_eq!(
            descriptor("conversation_bootstrap").lock_groups,
            HISTORY_LOCK
        );
        assert_eq!(
            descriptor("format_files").lock_groups,
            WORKSPACE_CONTENT_LOCK
        );
        assert_eq!(
            descriptor("git_restore").lock_groups,
            GIT_AND_WORKSPACE_LOCK
        );
        assert!(descriptor("git_status").log_operation);
        assert!(!descriptor("read_file").log_operation);
        assert!(descriptor("read_file").workspace_selector);
        assert!(descriptor("git_commit").workspace_selector);
        assert!(descriptor("exec_command").workspace_selector);
        assert!(!descriptor("history_session_checkpoint").workspace_selector);
    }

    #[test]
    fn aliases_and_argument_sensitive_policies_share_one_contract() {
        assert_eq!(descriptor("edit_file"), descriptor("edit"));
        assert_eq!(descriptor("edit_many"), descriptor("edit"));
        assert!(requires_write_baseline("edit", &serde_json::json!({})));
        assert!(!requires_write_baseline(
            "edit",
            &serde_json::json!({"dry_run": true})
        ));
        assert!(!request_mutates(
            "format_files",
            &serde_json::json!({"mode": "check"})
        ));
        assert!(request_mutates(
            "format_files",
            &serde_json::json!({"mode": "apply"})
        ));
    }
}
